### ISSUE TITLE
Discrepancy in ops.correlate for tensorflow

### ISSUE THREAD

**Issue by wesselvannierop** [keras-team/keras#21617]
For the tensorflow backend, `keras.ops.correlate` does not match numpy outputs in mode='same' and mode='full'. The other backends and mode='valid' work fine.

```python
import os

os.environ["KERAS_BACKEND"] = "tensorflow"

import numpy as np
from keras import ops

x = np.array([1, 3, 5])
y = np.array([7, 9])

mode = "full"
print("Mode: ", mode)
print("numpy: ", np.correlate(x, y, mode=mode))
print("tensorflow: ", ops.convert_to_numpy(ops.correlate(x, y, mode=mode)))

mode = "same"
print("Mode: ", mode)
print("numpy: ", np.correlate(x, y, mode=mode))
print("tensorflow: ", ops.convert_to_numpy(ops.correlate(x, y, mode=mode)))

mode = "valid"
print("Mode: ", mode)
print("numpy: ", np.correlate(x, y, mode=mode))
print("tensorflow: ", ops.convert_to_numpy(ops.correlate(x, y, mode=mode)))
```

```bash
Mode:  full
numpy:  [ 9 34 66 35]
tensorflow:  [34. 66. 35.  0.]
Mode:  same
numpy:  [ 9 34 66]
tensorflow:  [34. 66. 35.]
Mode:  valid
numpy:  [34 66]
tensorflow:  [34. 66.]
```

**Comment by wesselvannierop** [comment]
Besides, for jax, `jnp.correlate` supports complex values, while in torch and tensorflow it seems to cast to float.

I have an implementation that fixes both issues:

```python
def correlate(x, y, mode="full"):
    """
    Complex correlation via splitting real and imaginary parts.
    Equivalent to np.correlate(x, y, mode).

    NOTE: this function exists because tensorflow does not support complex correlation.
    NOTE: tensorflow also handles padding differently than numpy, so we manually pad the input.

    Args:
        x: np.ndarray (complex or real)
        y: np.ndarray (complex or real)
        mode: "full", "valid", or "same"
    """
    x = ops.convert_to_tensor(x)
    y = ops.convert_to_tensor(y)

    is_complex = "complex" in ops.dtype(x) or "complex" in ops.dtype(y)

    # Split into real and imaginary
    xr, xi = ops.real(x), ops.imag(x)
    yr, yi = ops.real(y), ops.imag(y)

    # Pad to do full correlation
    pad_left = ops.shape(y)[0] - 1
    pad_right = ops.shape(y)[0] - 1
    xr = ops.pad(xr, [[pad_left, pad_right]])
    xi = ops.pad(xi, [[pad_left, pad_right]])

    # Correlation: sum over x[n] * conj(y[n+k])
    rr = ops.correlate(xr, yr, mode="valid")
    ii = ops.correlate(xi, yi, mode="valid")
    ri = ops.correlate(xr, yi, mode="valid")
    ir = ops.correlate(xi, yr, mode="valid")

    real_part = rr + ii
    imag_part = ir - ri

    real_part = ops.cast(real_part, "complex64")
    imag_part = ops.cast(imag_part, "complex64")

    complex_tensor = real_part + 1j * imag_part

    # Extract relevant part based on mode
    full_length = ops.shape(real_part)[0]
    x_len = ops.shape(x)[0]
    y_len = ops.shape(y)[0]

    if mode == "same":
        # Return output of length max(M, N)
        target_len = ops.maximum(x_len, y_len)
        start = ops.floor((full_length - target_len) / 2)
        start = ops.cast(start, "int32")
        end = start + target_len
        complex_tensor = complex_tensor[start:end]
    elif mode == "valid":
        # Return output of length max(M, N) - min(M, N) + 1
        target_len = ops.maximum(x_len, y_len) - ops.minimum(x_len, y_len) + 1
        start = ops.ceil((full_length - target_len) / 2)
        start = ops.cast(start, "int32")
        end = start + target_len
        complex_tensor = complex_tensor[start:end]
    # For "full" mode, use the entire result (no slicing needed)

    if is_complex:
        return complex_tensor
    else:
        return ops.real(complex_tensor)
```

```bash
Mode:  full
numpy:  [ 9 34 66 35]
custom:  [ 9. 34. 66. 35.]
Mode:  same
numpy:  [ 9 34 66]
custom:  [ 9. 34. 66.]
Mode:  valid
numpy:  [34 66]
custom:  [34. 66.]
```

**Comment by mehtamansi29** [comment]
Hi @wesselvannierop - Please feel free to raise the PR regarding fixes the issue.

**Comment by monicadsong** [comment]
Hi @wesselvannierop --

The Keras team would like to learn more your Keras usage. Please thumbs up this comment if you are open to speaking with us and we'll reach out to you over email.

Thanks you so much!
Monica, Keras Product Manager

**Comment by LakshmiKalaKadali** [comment]
Hi @wesselvannierop,

I have verified the code([gist)](https://colab.sandbox.google.com/gist/LakshmiKalaKadali/bfe7eacff21e79acd1c80332f6229fc7/keras_21617.ipynb) with the latest keras 3.12.0 version.The bug has been resolved through this PR [#21778](https://github.com/keras-team/keras/pull/21778) and now the `keras.ops.correlate`  outputs matches with Numpy's.  

Thank You

**Comment by wesselvannierop** [comment]
Awesome! I also verified that it works, however, the fix is not yet in the latest `3.12.0` version, but only exists on `master` at the moment.

**Comment by wesselvannierop** [comment]
I also want to point toward the other issue I raised, which is that for complex `correlate` in the `torch` and `tensorflow` backend it casts to `float`. I am not sure if `keras` wants to support `complex` data-types? If not, maybe this should be mentioned in the docs.


```python
import os

os.environ["KERAS_BACKEND"] = "tensorflow" # or torch

from keras import ops
import numpy as np

for mode in ["full", "valid", "same"]:
    # Set random seed for reproducibility
    np.random.seed(42)

    # Test with complex vectors
    a_complex = (np.random.randn(8) + 1j * np.random.randn(8)).astype(np.complex64)
    v_complex = (np.random.randn(5) + 1j * np.random.randn(5)).astype(np.complex64)

    result_complex = ops.correlate(a_complex, v_complex, mode=mode)
    expected_complex = np.correlate(a_complex, v_complex, mode=mode)

    np.testing.assert_allclose(result_complex, expected_complex, rtol=1e-5, atol=1e-5)  # fails in torch and tensorflow!
```

**Comment by LakshmiKalaKadali** [comment]
Hi @wesselvannierop,

That's a good observation. Currently, Keras ops like `ops.correlate` don’t fully support complex dtypes `complex64` and `complex128` on the TensorFlow or PyTorch backends. In these cases, the tensors are implicitly cast to float32, which causes the imaginary components to be dropped.
This is a known limitation and also relates to [PR #19872](https://github.com/keras-team/keras/pull/19872), which begins improving complex number handling in Keras ops.
At the moment, Keras does not officially guarantee complex tensor support across all backends. We’d be happy to help guide you if you’d like to open a  PR.

Thank You

**Comment by wesselvannierop** [comment]
Okay, thanks for the response! In that case I'd be happy to close this issue, as the main problem has been addressed in #21778.

**Comment by PES2UG23CS205** [comment]
Hi all, thanks for the clarification!

I understand that the main TensorFlow padding/mode discrepancy has already been fixed in #21778.

I’d be happy to contribute by adding clear documentation around the current limitation of `keras.ops.correlate` with complex dtypes on the TensorFlow and PyTorch backends (where complex inputs are cast to float and the imaginary part is dropped), so users are aware of this behavior.

If that sounds good, I’ll open a small PR updating the API docs accordingly.

**Comment by wesselvannierop** [comment]
Great, it seems that everything I raised in this issue has been addressed and is now also available in v3.13.0. Thanks for the contributions!
